File: src/main/java/edu/cornell/cs/apl/viaduct/parsing/Parser.java
package edu.cornell.cs.apl.viaduct.parsing;

import edu.cornell.cs.apl.viaduct.syntax.surface.ProgramNode;
import java_cup.runtime.ComplexSymbolFactory;

class Parser {
  private Parser() {}

  /** Parses the source file and returns the AST. */
  static ProgramNode parse(SourceFile file) throws Exception {
    ComplexSymbolFactory symbolFactory = new ComplexSymbolFactory();
    ImpLexer scanner = new ImpLexer(file, symbolFactory);
    ImpParser parser = new ImpParser(scanner, symbolFactory);
    return (ProgramNode) parser.parse().value;
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/parsing/Parser.kt
package edu.cornell.cs.apl.viaduct.parsing

import edu.cornell.cs.apl.viaduct.syntax.surface.ProgramNode

/** Parses this string and returns the AST. */
fun String.parse(): ProgramNode {
    return SourceFile.from("<string>", this).parse()
}

/** Parses this source file and returns the AST. */
fun SourceFile.parse(): ProgramNode {
    return Parser.parse(this)
}
